pkg/workflow/providers/oam/apply.go

```diff
@@ -117,6 +129,54 @@ func (p *provider) ApplyComponent(ctx wfContext.Context, v *value.Value, act wfT
 	return nil
 }
 
+// ApplyComponent apply component.
+func (p *provider) ApplyComponent(ctx wfContext.Context, v *value.Value, act wfTypes.Action) error {
+	return p.applyComponent(ctx, v, act, nil)
+}
+
+// ApplyComponents apply components in parallel.
+func (p *provider) ApplyComponents(ctx wfContext.Context, v *value.Value, act wfTypes.Action) error {
+	components, err := v.LookupValue("components")
+	if err != nil {
+		return err
+	}
+	parallelism, err := v.GetInt64("parallelism")
+	if err != nil {
+		return err
+	}
+	if parallelism <= 0 {
+		return errors.Errorf("parallelism cannot be smaller than 1")
+	}
+	mu := &sync.Mutex{}
+	var wg sync.WaitGroup
+	ch := make(chan struct{}, parallelism)
+	var errs errors2.ErrorList
+	err = components.StepByFields(func(name string, in *value.Value) (bool, error) {
+		ch <- struct{}{}
+		wg.Add(1)
+		go func(_name string, _in *value.Value) {
+			defer func() {
+				wg.Done()
+				<-ch
+			}()
+			if err := p.applyComponent(ctx, _in, act, mu); err != nil {
+				mu.Lock()
```

Comment:
Why do we need a lock here?

Reply:
It is not thread safe. Without lock, data race could happen.